nat/src/main/java/org/hyperledger/besu/nat/NatService.java

```diff
@@ -0,0 +1,172 @@
+        LOG.warn("Caught exception while trying to stop the system or service", e);
+      }
+    }
+  }
+
+  /**
+   * Returns a {@link Optional} wrapping the advertised IP address.
+   *
+   * @return The advertised IP address wrapped in a {@link Optional}. Empty if
+   *     `isNatExternalIpUsageEnabled` is false
+   */
+  public Optional<String> queryExternalIPAddress() {
+    if (isNatEnvironment()) {
+      try {
+        final NatManager natSystem = getNatManager().orElseThrow();
+        LOG.info(
+            "Waiting for up to {} seconds to detect external IP address...",
+            NatManager.TIMEOUT_SECONDS);
+        return Optional.of(
+            natSystem.queryExternalIPAddress().get(NatManager.TIMEOUT_SECONDS, TimeUnit.SECONDS));
+
+      } catch (Exception e) {
+        LOG.warn(
+            "Caught exception while trying to query NAT external IP address (ignoring): {}", e);
+      }
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * Returns a {@link Optional} wrapping the local IP address.
+   *
+   * @return The local IP address wrapped in a {@link Optional}.
+   */
+  public Optional<String> queryLocalIPAddress() throws RuntimeException {
+    if (isNatEnvironment()) {
+      try {
+        final NatManager natSystem = getNatManager().orElseThrow();
+        LOG.info(
+            "Waiting for up to {} seconds to detect external IP address...",
+            NatManager.TIMEOUT_SECONDS);
+        return Optional.of(
+            natSystem.queryLocalIPAddress().get(NatManager.TIMEOUT_SECONDS, TimeUnit.SECONDS));
+      } catch (Exception e) {
+        LOG.warn("Caught exception while trying to query local IP address (ignoring): {}", e);
+      }
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * Returns the port mapping associated to the passed service type.
+   *
+   * @param serviceType The service type {@link NatServiceType}.
+   * @param networkProtocol The network protocol {@link NetworkProtocol}.
+   * @return The port mapping {@link NatPortMapping}
+   */
+  public Optional<NatPortMapping> getPortMapping(
+      final NatServiceType serviceType, final NetworkProtocol networkProtocol) {
+    if (isNatEnvironment()) {
+      try {
+        final NatManager natSystem = getNatManager().orElseThrow();
+        return Optional.of(natSystem.getPortMapping(serviceType, networkProtocol));
+      } catch (Exception e) {
+        LOG.warn("Caught exception while trying to query port mapping (ignoring): {}", e);
+      }
+    }
+    return Optional.empty();
+  }
+
+  @VisibleForTesting
+  void setNatSystem(final NatManager natManager) {
+    this.currentNatManager = Optional.of(natManager);
+  }
+
+  /** Initialize the current NatMethod. */
+  private Optional<NatManager> buildNatSystem(final NatMethod givenNatMethod) {
+    switch (givenNatMethod) {
+      case UPNP:
+        return Optional.of(new UpnpNatManager());
+      case NONE:
+      default:
+        return Optional.empty();
+    }
+  }
```

Comment:
What do you think about this class' constructor taking a `NatManager` instead of the enum in the constructor? The enum field can live in the manager, assigned again here, and we achieve more dependency injection. It would also remove the need for a `setNatSystem` method just for testing.

Reply:
In fact I just made the change and I added a builder for NatService